### PR TITLE
Desktop: Fixes #13903: Avoid OOM when printing notes with large attachment links

### LINKED ISSUE #13903: Printing a note crashes Joplin when it contains an internal resource link (:/resource-id)
### Operating system

Windows

### Joplin version

3.5.9

### Desktop version info

Joplin 3.5.9 (prod, win32)

Device: win32, AMD Ryzen 7 5825U with Radeon Graphics         
Client ID: 08752a84e28945c787628e853e3b2a64
Sync Version: 3
Profile Version: 49
Keychain Supported: Yes
Alternative instance ID: -

Revision: 4afac41

Backup: 1.4.3
Freehand Drawing: 3.1.0
Hotfolder: 1.2.2

### Current behaviour

Environment

Joplin Desktop 3.4.12

Windows 11 (fully up to date)

Occurs on multiple Windows systems

Plugins disabled / Safe Mode tested

Description

Printing a note in Joplin causes a renderer crash (Renderer process gone) when the note contains a Joplin internal resource link of the form:

[Filename.zip](:/resource-id)

The crash occurs consistently in normal mode.
Printing works if the line is removed or replaced with plain text.

Safe Mode allows printing, suggesting a renderer / print pipeline issue rather than data corruption.

Steps to Reproduce

Create or open a note

Add an attachment and keep the default Joplin resource link, e.g.:

[Games Windows7Games_for_Windows_11_10_8.zip](:/0cc4f305e4d34ae8a054ac46704f3362)

File → Print

Expected Result

The note prints successfully, or the resource link is ignored during printing.

Actual Result

Joplin crashes with:

An error occurred: Renderer process gone.

Notes

Other notes without :/resource-id links print correctly

Reinstalling Joplin and resetting profile does not resolve the issue

Replacing the resource link with plain text or an external file:// or https:// link prevents the crash

Issue appeared after recent Windows updates (possibly Electron/Chromium related)

Workaround

Remove or replace :/resource-id links before printing, or print in Safe Mode.

### Expected behaviour

Print properly as other notes have

### Logs

_No response_

### PR DESCRIPTION
**Pull request by Rygaa**
## Summary

Printing could crash the renderer with an out-of-memory error when a note contained links to large non-image resources (e.g. archives). Packing HTML for print/export inlined those `<a href="...">` targets as base64 `data:` URLs, which forced the whole file through encoding and blew memory.

## Solution

- Remove anchor-tag inlining from `packToString` so links keep their original `href` (no `data:` URL for attachments). Images, stylesheets, and scripts are unchanged.
- Normalize CRLF to LF on the packed HTML string so output is consistent regardless of fixture or platform newlines; the integration test builds the expected document with `.join('\n')` so it stays stable on Windows/Git Bash.

This matches the agreed product trade-off: standalone HTML will no longer embed those files as `data:` links, but printing no longer OOMs.

## Testing

Create a new note or open an existing one.
Attach a large ZIP file to the note.
In Joplin, click File → Print.
Joplin should no longer crash.

### CODE REVIEW

File: packages/htmlpack/packToString.ts [hunk 1]
```diff
@@ -166,19 +166,6 @@ const packToString = async (baseDir: string, inputFileText: string, fs: FileApi)
...
-		const filePath = `${baseDir}/${href}`;
-		if (!await fs.exists(filePath)) return null;
-
-		const modAttrs = { ...attrs };
-		modAttrs.href = await readFileDataUriSafe(filePath);
-		modAttrs.download = basename(href);
-		return `<a ${attributesHtml(modAttrs)}>`;
-	};
-
 	const output: string[] = [];
```
Comment (laurent22) [inline packages/htmlpack/packToString.ts]: But this is used in other contexts, isn't it? The idea of htmlpack is that all resources are included in the same file - it's used to export notes to HTML files too with all resources.

If that's a problem for PDF files, we can disable it for this, but it should still be enabled for other cases

Comment (Rygaa) [reply packages/htmlpack/packToString.ts]: hmmm it did not crash only in the html directory export but it did crash in html export.

```
## Crashed

- Print
- Export PDF
- Note export -> HTML File
- Sidebar notebook export -> HTML File
- Programmatic exportFolders -> HTML file target
- Programmatic exportFolders -> HTML directory target

## Worked

- Note export -> HTML Directory
- Sidebar notebook export -> HTML Directory
- Main menu -> Export all -> HTML Directory
- Programmatic exportNotes -> HTML directory target

```

Comment (laurent22) [reply packages/htmlpack/packToString.ts]: So we need a more nuanced fix then. The idea was to disable embedding resources for PDF. We haven't discussed disabling it for HTML files and we don't really want this. What are the options?

Comment (Rygaa) [reply packages/htmlpack/packToString.ts]: The issue is caused by readFile(path, 'base64'). We should keep embedding for HTML file exports, but switch that part to use streaming or chunking instead of loading the entire file into memory at once.

Comment (Rygaa) [reply packages/htmlpack/packToString.ts]: The task took a bit more time than I expected because I first tried to keep output.push while adding streaming, so the change would stay small and safe. But that caused another oom problem, because output.push was still keeping a very large string in memory. For the PDF part, I also had to update app-desktop so it can know if the call is for a normal HTML export or for a PDF export, because these two cases need different behavior.

Comment (laurent22) [reply packages/htmlpack/packToString.ts]: That makes sense, thanks @Rygaa. There are some CodeRabbit comments to check now

**Comment by Rygaa** [comment]
@coderabbitai review

**Comment by Rygaa** [comment]
**My concern with my current fix**


Before, processedResult was simple:
empty string meant "nothing special happened, so write the normal tag"
a non-empty string meant "this tag was transformed, so write that HTML instead"

Now, because processAnchorTag can stream directly to write(...) instead of building one huge string, processedResult needs to represent one more case:
null means "nothing special happened, write the normal tag"
a non-empty string means "write this transformed HTML"
'' means "this tag was already handled by streaming, so do not write anything again here"

My conclusion: 
I find the new 3 states a bit confusing and also having processAnchorTag create some edge case return by itself makes the code a bit awkward

**Comment by Rygaa** [comment]
https://github.com/user-attachments/assets/931d42a2-29c7-473b-b711-b079d5bcbf2e

File: packages/htmlpack/packToString.ts [hunk 2]
```diff
@@ -50,14 +50,17 @@ const isSelfClosingTag = (tagName: string) => {
...
 export type FileApi = {
 	exists(path: string): Promise<boolean>;
 	readFileText(path: string): Promise<string>;
 	readFileDataUri(path: string): Promise<string>;
+	streamFileDataUri?(path: string, onChunk: FileApiChunkCallback): Promise<void>;
 };
 
 // packToString should be able to run in React Native -- don't use fs-extra.
-const packToString = async (baseDir: string, inputFileText: string, fs: FileApi) => {
+const packToString = async (baseDir: string, inputFileText: string, fs: FileApi, write: FileApiChunkCallback) => {
```
Comment (laurent22) [inline packages/htmlpack/packToString.ts]: We're already passing a `FileApi` to this function - is it not possible to use this to implement the file chunking functions?

Also this function is called `packToString`, which is no longer what it does. Please update the name

File: packages/app-desktop/InteropServiceHelper.ts [hunk 3]
```diff
@@ -58,6 +58,7 @@ export default class InteropServiceHelper {
 			const exportOptions = {
 				customCss: options.customCss ? options.customCss : '',
 				plugins: options.plugins,
+				packIntoSingleFile: false,
```
Comment (laurent22) [inline packages/app-desktop/InteropServiceHelper.ts]: Should that be here? I see that in InteropService_Exporter_Html.ts, packIntoSingleFile defaults to `true` but now it seems you make it default to `false` but from somewhere else. Also why do we need this now that you've implemented the streaming API?

Comment (Rygaa) [reply packages/app-desktop/InteropServiceHelper.ts]: The confusion is that I originally focused on large zip files, where most PDF viewers do not support opening a zip from a data: link. So I set packIntoSingleFile: false to skip packing for the PDF/print path, without considering that other file types can still be usefully embedded in a PDF. So yes, packIntoSingleFile should be removed so PDF can also embed other file types.

Laurent confirmed the following -> For this, I just want to confirm: most PDF viewers don't really support opening a zip from a data: link, even when the zip is small, right? I want to be sure before arguing for a different fix.

The remaining problem is that we should still ignore zip files in a PDF even with streaming working, having a zip inside a PDF is useless. Should I add logic to skip packing archive types (zip, 7z, ...) for PDF?

File: packages/htmlpack/packToString.ts (hunk 2 above)
Comment (Rygaa) [reply packages/htmlpack/packToString.ts]: fixed

File: packages/app-desktop/InteropServiceHelper.ts (hunk 3 above)
Comment (laurent22) [reply packages/app-desktop/InteropServiceHelper.ts]: So as discussed before we don't embed anything in the pdf, just images

**Comment by laurent22** [comment]
@Rygaa please address code rabbit comments

File: packages/htmlpack/package.json [hunk 4]
```diff
@@ -9,9 +9,9 @@
       "types": "./index.ts",
       "default": "./dist/index.js"
     },
-    "./packToString": {
-      "types": "./packToString.ts",
-      "default": "./dist/packToString.js"
+    "./packToWriter": {
+      "types": "./packToWriter.ts",
+      "default": "./dist/packToWriter.js"
```
[the remark this answers is not in the document]
Comment (Rygaa) [reply packages/htmlpack/package.json:18]: It will be pushed in the next commit

[the remark this answers is not in the document]
Comment (Rygaa) [reply packages/htmlpack/package.json:18]: Are you sure you are not wrong in your review because:
All Joplin packages use the same version number together.
The only file that used packToString was already updated in this PR.
No outside code uses this package directly.

File: packages/htmlpack/packToWriter.ts [hunk 5]
```diff
@@ -50,14 +50,18 @@ const isSelfClosingTag = (tagName: string) => {
...
 	readFileText(path: string): Promise<string>;
 	readFileDataUri(path: string): Promise<string>;
+	streamFileDataUri?(path: string, onChunk: FileApiChunkCallback): Promise<void>;
+	writeChunk(chunk: string): void | Promise<void>;
 };
 
-// packToString should be able to run in React Native -- don't use fs-extra.
-const packToString = async (baseDir: string, inputFileText: string, fs: FileApi) => {
+// packToWriter should be able to run in React Native -- don't use fs-extra.
+const packToWriter = async (baseDir: string, inputFileText: string, fs: FileApi) => {
```
[the remark this answers is not in the document]
Comment (Rygaa) [reply packages/htmlpack/packToWriter.ts:64]: Before my fixes was the problem not present?

File: packages/htmlpack/packToWriter.ts [hunk 6]
```diff
@@ -166,20 +170,35 @@ const packToString = async (baseDir: string, inputFileText: string, fs: FileApi)
...
+
+		delete modAttrs.href;
+		await fs.writeChunk('<a href="');
+		if (fs.streamFileDataUri) {
+			await fs.streamFileDataUri(filePath, async (chunk) => { await fs.writeChunk(chunk); });
+		} else {
+			await fs.writeChunk(await readFileDataUriSafe(filePath));
+		}
+		await fs.writeChunk(`" ${attributesHtml(modAttrs)}>`);
+		return '';
```
[the remark this answers is not in the document]
Comment (Rygaa) [reply packages/htmlpack/packToWriter.ts:200]: The preexisting test was expecting those results. Confirm it's true
```
	test('should not throw an error on invalid resource paths', (async () => {
		const service = InteropService.instance();
		const folder1 = await Folder.save({ title: 'folder1' });
		await Note.save({ title: 'note1', parent_id: folder1.id, body: '[a link starts with slash](/)' });

		const filePath = `${exportDir()}/test.html`;

		await service.export({
			path: filePath,
			format: ExportModuleOutputFormat.Html,
			target: FileSystemItem.File,
		});

		const content = await readFile(filePath, 'utf-8');
		expect(content).toContain('<a data-from-md="" title="/" href="" download="">a link starts with slash</a>');
	}));

```

[the remark this answers is not in the document]
Comment (Rygaa) [reply packages/htmlpack/packToWriter.ts:200]: The OOM problem happens during the export, not in the browser. With streamFileDataUri, large files are read in small pieces (3 MB at a time) and written directly to disk. The full file is never in memory at once. That is the fix. The final HTML file will always be large when files are embedded that cannot be avoided


verify my claim

**Comment by Rygaa** [comment]
coderabbitai has been addressed. except this one [https://github.com/laurent22/joplin/pull/15026#discussion_r3081848541](url)

This problem is not specific to packToWriter. The same issue exists in many other places in the codebase that read and inject CSS or JS files. Fixing it only here would be inconsistent. This is already being tracked as a broader Windows issue in [.](https://github.com/laurent22/joplin/pull/14905) It needs a dedicated PR that addresses the issue as a whole.

File: packages/app-desktop/InteropServiceHelper.ts [hunk 7]
```diff
@@ -58,6 +58,7 @@ export default class InteropServiceHelper {
 			const exportOptions = {
 				customCss: options.customCss ? options.customCss : '',
 				plugins: options.plugins,
+				shouldEmbedOnlyImages: true,
```
Comment (laurent22) [inline packages/app-desktop/InteropServiceHelper.ts:61]: Is it possible to implement this feature without having this here? For PDF we only want to embed images, for everything else we want to embed everything. So if there is such a config option it should be much closer to wherever PDF it being created.

File: packages/htmlpack/package.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (laurent22) [reply packages/htmlpack/package.json:18]: I saw that @joplin/htmlpack was used by at least one other project. We can change internal functions but we shouldn't needlessly break external APIs

File: packages/app-desktop/InteropServiceHelper.ts (hunk 7 above)
Comment (Rygaa) [reply packages/app-desktop/InteropServiceHelper.ts:61]: exportNoteTo_() is only called from exportNoteToPdf() and printNote() both produce a PDF (as a file or via the print dialog). So shouldEmbedOnlyImages: true is already scoped to PDF only, there's no other path that calls this. It's also the only place in the code where we know the output is a PDF so it makes sense to set it here.

File: packages/htmlpack/package.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (Rygaa) [reply packages/htmlpack/package.json:18]: Within this repo, only /lib uses /htmlpack and that import was updated as part of this PR. For external usage you're right, I didn't confirm with you before making the change. I renamed it because you asked for it here [https://github.com/laurent22/joplin/pull/15026#discussion_r3078526639](url). Should I keep ./packToString as an alias in package.json to avoid breaking external consumers?

File: packages/app-desktop/InteropServiceHelper.ts (hunk 7 above)
Comment (laurent22) [reply packages/app-desktop/InteropServiceHelper.ts:61]: Right, I thought I was looking at InteropService. It's fine then, thank you for clarifying

File: packages/htmlpack/package.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (laurent22) [reply packages/htmlpack/package.json:18]: @Rygaa

[the remark this answers is not in the document]
Comment (Rygaa) [reply packages/htmlpack/package.json:18]: During the Thursday meeting, I discussed this with you, and you said we should wrap packToWriter to keep support for packToString for now. You also said this could avoid the OOM issue in the new packToString. I replied that I did not think this was possible, but I still tried your suggestion. After that, I showed the solution to Henry, and he agreed with me that this version can still cause OOM, because packToString still needs to keep the full result in memory at the end. He also said this is not very important for this issue, because people who want to fix it should use the new packToWriter instead.

Let me know if I misunderstood your suggestion and tried something different.

Clarification: this may improve the OOM situation a bit, because the old packToString stored all parts in output[] and then joined them at the end, while the new version builds the final result only once from chunks.join(''). But it still cannot fully avoid OOM, because packToString must return the whole packed string in memory.

File: packages/htmlpack/packToWriter.test.ts [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+			},
+			writeChunk: (chunk: string) => {
+				writtenChunks.push(chunk);
+			},
+		};
+
+		await packToWriter('/base', '<a href="big.bin">d</a>', fs);
+
+		expect(onChunkCalls).toBe(5);
+	});
```
Comment (laurent22) [inline packages/htmlpack/packToWriter.test.ts:43]: I don't really understand the point of this test. Any reason why you don't simply test what the function does? Create some temp dir/files, run packToWriter on it and check that the output is as expected?

Comment (Rygaa) [reply packages/htmlpack/packToWriter.test.ts:43]: This test checks behavior, not just final output. A normal output test could still pass even if packToWriter loaded the whole file into memory and wrote it only at the end. But the purpose of packToWriter is to avoid keeping large files in memory and prevent OOM errors, so this test makes sure each chunk is written immediately before the next one is read.

Comment (laurent22) [reply packages/htmlpack/packToWriter.test.ts:43]: Please test that the function actually does what it does. I'm not convinced that your test helps much since it implements a new fs interface just for the test so if the real one is faulty we won't know. But in any case, please test what the function actually does at a minimum

Comment (Rygaa) [reply packages/htmlpack/packToWriter.test.ts:43]: Then "packages/htmlpack/index.test.ts" should already be enough. The function in index.ts is just a small wrapper around packToWriter. So the existing tests in "index.test.ts" already cover that. Adding another test file that does the same thing would just be a duplicate.

Comment (laurent22) [reply packages/htmlpack/packToWriter.test.ts:43]: Ok I see and the new ability not to embed media is tested in the InteropService test, so I guess it's fine. I'd suggest to either clarify what the test here does or remove it

Comment (Rygaa) [reply packages/htmlpack/packToWriter.test.ts:43]: - I updated the test name and comments to make the purpose more clear. 
- I decided to keep it because the real behavior was already tested in index.test.ts and InteropService. The only remaining untested behavior is the OOM which this test cover by making sure the data is written in chunks

Comment (personalizedrefrigerator) [reply packages/htmlpack/packToWriter.test.ts:43]: Since `htmlpack` is published on NPM, I think it could make sense to have tests that directly test the public APIs (even if the function is indirectly covered elsewhere).

Additionally: Do any of the existing tests use input/output files large enough to produce multiple input/output chunks? If not, I think this test also makes sense as a way to ensure that multiple chunk cases are tested. **Edit**: The chunk size is 3 MB, so chunking behavior might not be triggered by the other `htmlpack` tests.
